Sources/StreamChat/APIClient/APIClient.swift

```diff
@@ -90,6 +128,46 @@ class APIClient {
             completion: completion
         )
     }
+    
+    /// Queues a failed request for executing later or failing after a timeout
+    func handleTokenExpirationError<Response: Decodable>(
+        urlRequest: URLRequest,
+        timeout: TimeInterval,
+        completion: @escaping (Result<Response, Error>) -> Void
+    ) {
+        requestsAccessQueue.async {
+            let item = RequestsQueueItem(
+                requestAction: { self.request(
+                    urlRequest: urlRequest,
+                    timeout: timeout,
+                    completion: completion
+                )
+                },
+                failureAction: { completion(.failure(ClientError.ExpiredToken())) }
+            )
+            
+            // The moment we queue a first request for execution later we also set a timeout for 60 seconds
+            // after which we fail all the queued requests
+            if self.requestsQueue.isEmpty {
+                self.requestsAccessQueue.asyncAfter(deadline: .now() + timeout) {
+                    self.requestsQueue.forEach { $0.failureAction() }
+                    self.requestsQueue = []
+                }
+            }
+            
+            self.requestsQueue.append(item)
+            
+            // Only initiate token refreshing once
+            guard !self.isRefreshingToken else { return }
```

Comment:
see previous comment about concurrent usage of self and `isRefreshingToken `. If needed the Atomic `propertyWrapper` can be really useful here.

```swift

/// ...
self.requestsQueue.append(item)

guard self._ isRefreshingToken.compareAndSwap(old: false, new: true) else {
   return
}

self.requestsAccessQueue.asyncAfter(deadline: .now() + timeout) {
   self.requestsQueue.forEach { $0.failureAction() }
   self.requestsQueue = []
}

self.tokenRefresher(...)
```